include/swift/Basic/Statistic.h

```diff
@@ -22,4 +25,68 @@
     ++FStat;                                                            \
   } while (0)
 
+// Helper class designed to consolidate reporting of LLVM statistics and timers
+// across swift compilations that typically invoke many drivers, each running
+// many frontends. Additionally collects some cheap "always-on" statistics,
+// beyond those that are (compile-time) parametrized by -DLLVM_ENABLE_STATS
+// (LLVM's stats are global and involve some amount of locking and mfences).
+//
+// Assumes it's given a process name and target name (the latter used as
+// decoration for its self-timer), and a directory to collect stats into, then:
+//
+//  - On construction:
+//    - Calls llvm::EnableStatistics(/*PrintOnExit=*/false)
+//    - Calls swift::enableCompilationTimers()
+//    - Starts an llvm::NamedRegionTimer for this process
+//
+//  - On destruction:
+//    - Add any standard always-enabled stats about the process as a whole
+//    - Opens $dir/stats-$timestamp-$name-$random.json for writing
+//    - Calls llvm::PrintStatisticsJSON(ostream) and/or its own writer
+//
+// Generally we make one of these per-process: either early in the life of the
+// driver, or early in the life of the frontend.
+
+namespace swift {
+namespace driver {
+class Compilation;
+class Job;
+}
+
+class UnifiedStatsReporter : public FrontendObserver {
+  struct AlwaysOnDriverCounters;
+  struct AlwaysOnFrontendCounters;
+
+  SmallString<128> Filename;
+  std::unique_ptr<llvm::NamedRegionTimer> Timer;
+  std::unique_ptr<AlwaysOnDriverCounters> DriverCounters;
+  std::unique_ptr<AlwaysOnFrontendCounters> FrontendCounters;
+
+  AlwaysOnDriverCounters &getDriverCounters();
+  AlwaysOnFrontendCounters &getFrontendCounters();
+  void publishAlwaysOnStatsToLLVM();
+  void printAlwaysOnStatsAndTimers(llvm::raw_ostream &OS);
+public:
+  UnifiedStatsReporter(StringRef ProgramName,
+                       StringRef TargetName,
+                       StringRef Directory);
+  ~UnifiedStatsReporter();
+
+  // Callbacks from Driver
+  void driverMarkNode(uint8_t DepKinds, bool IsCascading);
+  void driverCompilationJobFinished(const driver::Compilation &C,
+                                    const driver::Job *J,
+                                    bool Skipped);
+
+  // FrontendObserver methods
+  void parsedArgs(CompilerInvocation &invocation) override;
+  void configuredCompiler(CompilerInstance &instance) override;
+  void performedSemanticAnalysis(CompilerInstance &instance) override;
+  void performedSILGeneration(SILModule &module) override;
+  void performedSILDiagnostics(SILModule &module) override;
+  void performedSILOptimization(SILModule &module) override;
```

Comment:
These are inherited methods from the FrontendObserver, I'm not able to change their type signature (am I?).